### PR TITLE
feat: Custom Placement Config Dual Region Support

### LINKED ISSUE #1428: Add Support for Updated Dual-Region Bucket API
[Internal]
- For details and coordination: [go/gcs-dpe-dual-region-milestones-api-refactor](http://go/gcs-dpe-dual-region-milestones-api-refactor)
- Implementation Example: https://github.com/googleapis/nodejs-storage/pull/1977

### PR DESCRIPTION
**Pull request by sydney-munro**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/java-storage/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #1428  ☕️

Will follow up with sample update.

If you write sample code, please follow the [samples format](
https://github.com/GoogleCloudPlatform/java-docs-samples/blob/main/SAMPLE_FORMAT.md).

### CODE REVIEW

**Review by danielbankhead — APPROVED** [review]
Thanks! LGTM

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 1]
```diff
@@ -103,6 +103,7 @@ public com.google.api.services.storage.model.Bucket apply(BucketInfo bucketInfo)
   private final IamConfiguration iamConfiguration;
   private final String locationType;
   private final Logging logging;
+  private final CustomPlacementConfig customPlacementConfig;
```
Comment (BenWhitehead) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: How does custom placement differ from `location` above?

Should we refactor away location to always use this multi-region type thing with a single element?

**Review by BenWhitehead — CHANGES_REQUESTED** [review]
Looks like some tests are failing and need to be fixed before merging.

I had one question above for BucketInfo.

@danielbankhead Do you know if this same change is going to be made for the gRPC protos?

We need to keep the protos in mind when designing our new model classes.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java (hunk 1 above)
Comment (danielbankhead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: Good catch, both `location` and `customPlacementConfig` will need to be specified for (custom) dual-regions. Here's an example:
- `location` = 'US'
- `customPlacementConfig` = 'US-EAST-1' and 'US-WEST-1'

**Comment by danielbankhead** [comment]
@sydney-munro, @BenWhitehead this PR should be testable/verifiable after a config push deployment (ETA July 1).

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java (hunk 1 above)
Comment (danielbankhead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: This should be made for the gRPC protos as well.

Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: @sydney-munro and @frankyn 

Given the information from Dan above: Do we want to introduce an actual higher level data type which can model the `location` vs `location + customPlacementConfig`, where our BucketInfo would then use this new type? Or, do we hope any possible error message from the backend helps folks figure out they need to be set together?

Comment (sydney-munro) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: I spoke with Dan offline and the backend should be handling this. I think the main reason to not do this is that using location does not necessarily mean you need customPlacementConfig.

Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: I was thinking of it from the other direction. If I'm using custom placement config I would guess that I choose either `location` or `custom placement config`, not choose either `location` or `location and custom placement config`.

Comment (sydney-munro) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: As I was working on coding this I do have a concern around this as a breaking change, do we want people that don't care about multiple regions that are using location to have to worry about changing their code just to be able to use the latest client?

Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java:106]: sync'd offline. we're going to go ahead with the initial proposal, deferring this new type to a later time if we need to.

**Review by BenWhitehead — APPROVED** [review]
Feel free to merge when tests are all passing.

**Pull request merged**